core/lib/zksync_core/src/state_keeper/keeper.rs

```diff
@@ -234,6 +226,59 @@ impl ZkSyncStateKeeper {
         Err(Error::Canceled)
     }
 
+    async fn load_protocol_upgrade_tx(
+        &mut self,
+        pending_miniblocks: &[MiniblockExecutionData],
+        protocol_version: ProtocolVersionId,
+        first_batch_in_shared_bridge: bool,
+    ) -> Result<Option<ProtocolUpgradeTx>, Error> {
+        let previous_batch_protocol_version =
+            self.io.load_previous_batch_version_id().await.unwrap();
+
+        let version_changed = protocol_version != previous_batch_protocol_version;
+        let mut protocol_upgrade_tx = self.io.load_upgrade_tx(protocol_version).await;
+
+        // After the Shared Bridge is integrated,
+        // there has to be a setChainId upgrade transaction after the chain genesis.
+        // It has to be the first transaction of the first batch.
+        // If it's not present, we have to wait for it to be picked up by the event watcher.
+        // The setChainId upgrade does not bump the protocol version, rather attaches an upgrade
+        // transaction to the genesis protocol version.
+        while first_batch_in_shared_bridge && protocol_upgrade_tx.is_none() {
+            tracing::info!("Waiting for setChainId upgrade transaction to be picked up");
+            if self.is_canceled() {
+                return Err(Error::Canceled);
+            }
+            tokio::time::sleep(Duration::from_secs(1)).await;
+            protocol_upgrade_tx = self.io.load_upgrade_tx(protocol_version).await;
+        }
+
+        let protocol_upgrade_tx =
+            if pending_miniblocks.is_empty() && (version_changed || first_batch_in_shared_bridge) {
+                tracing::info!("We have upgrade tx to be executed in empty miniblock");
```

Comment:
this log might not be correct (if version was changed, but for some reason the protocoL_upgrade_tx is still 'none')

Reply:
In this case it cannot be `None`. If `first_batch_in_shared_bridge` is `true`, we waited for it to be `Some` in the while-loop before. If `version_changed` is true, then `eth_watch` saved the upgrade transaction together with the new protocol version. But we can put an additional assert there just to be sure.